File: app/frontend/lib/clientMetrics.ts
// Bridges a small, closed set of browser-observed events (see
// app/services/client_metrics.rb) into real Prometheus counters. Several of
// EPIC-392's amplification signals -- an entity-store patch landing, a
// revision-gap recovery firing, a hidden tab suppressing a fetch -- only
// happen in the browser, and the backend metrics library has no way to see
// them directly. This is the one narrow bridge: callers record an
// occurrence, increments batch in memory, and the batch posts to
// /api/v1/app/client_metrics (mirroring performanceMarkers.ts's batching
// shape) instead of one request per occurrence.
import { readInitialBootstrap } from "../api/bootstrap"

export type ClientMetricResource =
  | "job" | "workflow" | "step" | "run" | "chat" | "dashboard" | "epic" | "repository" | "notification" | "unknown"

export type ClientMetricName = "entity_patch_applications" | "revision_gap_recoveries" | "hidden_tab_suppressed_fetches"

export type ClientMetricVisibilityState = "visible" | "hidden" | "unknown"

type QueuedIncrement = { name: ClientMetricName; resource: ClientMetricResource; visibility_state?: ClientMetricVisibilityState; by: number }

const FLUSH_DELAY_MS = 10_000
const FLUSH_SIZE = 50

// The backend's own closed resource enum (ClientMetrics::RESOURCES);
// anything else already normalizes to "unknown" via resourceTagFor below, but
// keeping the two lists in sync is what makes that mapping meaningful.
const KNOWN_RESOURCES: readonly string[] = [
  "job", "workflow", "step", "run", "chat", "dashboard", "epic", "repository", "notification"
]

let queue = new Map<string, QueuedIncrement>()
let flushTimer: ReturnType<typeof setTimeout> | null = null
let lifecycleListenersInstalled = false
let lifecycleCleanups: Array<() => void> = []

export function recordClientMetric(
  name: ClientMetricName,
  resource: ClientMetricResource = "unknown",
  visibilityState?: ClientMetricVisibilityState
): void {
  const key = `${name}:${resource}:${visibilityState ?? ""}`
  const existing = queue.get(key)
  if (existing) {
    existing.by += 1
  } else {
    queue.set(key, { name, resource, visibility_state: visibilityState, by: 1 })
  }
  scheduleFlush()
}

export function currentVisibilityState(): ClientMetricVisibilityState {
  if (typeof document === "undefined" || typeof document.visibilityState !== "string") return "unknown"
  return document.visibilityState === "visible" || document.visibilityState === "hidden" ? document.visibilityState : "unknown"
}

// Maps a query-key root ("jobs", "chats", ...) or an application-event
// `resource` ("job", "chat", ...) onto the closed tag vocabulary the backend
// accepts, so a plural query-key root and a singular event resource both land
// on the same series instead of silently becoming "unknown".
export function resourceTagFor(value: string | null | undefined): ClientMetricResource {
  if (!value) return "unknown"

  const singular = value.endsWith("s") ? value.slice(0, -1) : value
  return (KNOWN_RESOURCES.includes(singular) ? singular : "unknown") as ClientMetricResource
}

export function flushClientMetricsQueue(options: { beacon?: boolean } = {}): void {
  flushQueue(options)
}

export function resetClientMetricsForTest(): void {
  queue = new Map()
  if (flushTimer != null) {
    clearTimeout(flushTimer)
    flushTimer = null
  }
  lifecycleCleanups.forEach((cleanup) => cleanup())
  lifecycleCleanups = []
  lifecycleListenersInstalled = false
}

function scheduleFlush(): void {
  ensureLifecycleListeners()
  if (queue.size >= FLUSH_SIZE) {
    flushQueue()
    return
  }
  if (flushTimer != null) return
  flushTimer = setTimeout(() => flushQueue(), FLUSH_DELAY_MS)
}

function flushQueue(options: { beacon?: boolean } = {}): void {
  if (flushTimer != null) {
    clearTimeout(flushTimer)
    flushTimer = null
  }
  if (queue.size === 0) return

  const batch = Array.from(queue.values())
  queue = new Map()
  postClientMetrics(batch, options)
}

function postClientMetrics(clientMetrics: QueuedIncrement[], options: { beacon?: boolean }): void {
  const csrfToken = currentCsrfToken()
  const body = JSON.stringify({ client_metrics: clientMetrics, authenticity_token: csrfToken })

  if (options.beacon && typeof navigator !== "undefined" && typeof navigator.sendBeacon === "function") {
    const blob = new Blob([ body ], { type: "application/json" })
    if (navigator.sendBeacon("/api/v1/app/client_metrics", blob)) return
  }

  void fetch("/api/v1/app/client_metrics", {
    method: "POST",
    credentials: "same-origin",
    keepalive: true,
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      ...(csrfToken ? { "X-CSRF-Token": csrfToken } : {})
    },
    body
  }).catch(() => {})
}

function currentCsrfToken(): string | undefined {
  if (typeof document === "undefined") return readInitialBootstrap()?.csrf_token
  return readInitialBootstrap()?.csrf_token || document.querySelector<HTMLMetaElement>("meta[name='csrf-token']")?.content
}

function ensureLifecycleListeners(): void {
  if (lifecycleListenersInstalled) return
  if (typeof document === "undefined") return
  lifecycleListenersInstalled = true

  const flushWithBeacon = () => flushQueue({ beacon: true })
  const onVisibilityChange = () => {
    if (document.visibilityState === "hidden") flushWithBeacon()
  }
  document.addEventListener("visibilitychange", onVisibilityChange)
  lifecycleCleanups.push(() => document.removeEventListener("visibilitychange", onVisibilityChange))

  if (typeof window !== "undefined") {
    window.addEventListener("pagehide", flushWithBeacon)
    lifecycleCleanups.push(() => window.removeEventListener("pagehide", flushWithBeacon))
  }
}

File: app/services/client_metrics.rb
# Turns a small, closed set of frontend-observed events into real Prometheus
# counters. Several of the signals worth measuring --
# an entity-store patch actually landing, a revision-gap recovery firing, a
# hidden tab suppressing a fetch -- only happen in the browser, and this
# metrics library has no way to reach into a browser process. This is the
# one narrow bridge: app/frontend/lib/clientMetrics.ts batches increments and
# posts them to Api::V1::App::ClientMetricsController, which calls #record.
#
# Both `name` and `resource` are closed enums, not free strings, for the same
# reason Metrics::ProductUsage's FEATURES is closed: this is a public,
# authenticated HTTP boundary, so an unrecognized value is silently dropped
# in every environment rather than raised -- unlike ProductUsage, which is
# only ever called from trusted server-side code and can afford to raise in
# development/test. `by` is clamped so a single misbehaving tab cannot skew
# a counter by an implausible amount in one request.
module ClientMetrics
  REPORTABLE = {
    "entity_patch_applications" => :syrus_client_entity_patch_applications_total,
    "revision_gap_recoveries" => :syrus_client_revision_gap_recoveries_total,
    "hidden_tab_suppressed_fetches" => :syrus_client_hidden_tab_suppressed_fetches_total
  }.freeze

  RESOURCES = %w[job workflow step run chat dashboard epic repository notification unknown].freeze
  VISIBILITY_STATES = %w[visible hidden unknown].freeze
  MAX_BY = 1_000

  def self.declare_metrics!
    Syrus::Metrics.declare do
      counter :client_entity_patch_applications_total, tags: %i[resource visibility_state],
              comment: "Browser entity-store patches applied directly from an application event, by resource and " \
                       "the tab's visibility state when it applied -- reported by the frontend"
      counter :client_revision_gap_recoveries_total, tags: %i[resource],
              comment: "Browser-detected application-event sequence gaps recovered with a targeted refetch, by " \
                       "resource -- reported by the frontend"
      counter :client_hidden_tab_suppressed_fetches_total, tags: %i[resource],
              comment: "Refetches a hidden browser tab deferred instead of running immediately, by resource -- " \
                       "reported by the frontend"
    end
  end
  declare_metrics!

  def self.record(name:, resource: nil, visibility_state: nil, by: 1)
    metric = REPORTABLE[name.to_s]
    return unless metric

    tags = { resource: RESOURCES.include?(resource.to_s) ? resource.to_s : "unknown" }
    tags[:visibility_state] = VISIBILITY_STATES.include?(visibility_state.to_s) ? visibility_state.to_s : "unknown" if name.to_s == "entity_patch_applications"
    Syrus::Metrics.counter(metric).increment(tags: tags, by: clamp_by(by))
  end

  def self.clamp_by(by)
    Integer(by).clamp(1, MAX_BY)
  rescue ArgumentError, TypeError
    1
  end
  private_class_method :clamp_by
end
